### PR TITLE
[flink] Add CompactManifestProcedure to enable compact manifest manually

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

Come up with a flink procedure, which could compact manifest manually.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java [hunk 1]
```diff
@@ -1104,6 +1064,97 @@ CommitResult tryCommitOnce(
...
+        List<ManifestFileMeta> mergeBeforeManifests =
+                manifestList.readDataManifests(latestSnapshot);
+
+        List<ManifestFileMeta> mergeAfterManifests =
+                ManifestFileMerger.merge(
+                        mergeBeforeManifests,
+                        manifestFile,
+                        manifestTargetSize.getBytes(),
+                        1,
+                        manifestFullCompactionSize.getBytes(),
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:1083]: should be 1?

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java [hunk 2]
```diff
@@ -1104,6 +1064,97 @@ CommitResult tryCommitOnce(
...
+                        0L,
+                        latestSnapshot.watermark(),
+                        latestSnapshot.statistics());
+
+        Path newSnapshotPath =
+                branchName.equals(DEFAULT_MAIN_BRANCH)
+                        ? snapshotManager.snapshotPath(newSnapshot.id())
+                        : snapshotManager.copyWithBranch(branchName).snapshotPath(newSnapshot.id());
+
+        commitSnapshotImpl(newSnapshot, newSnapshotPath);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:1115]: why not retry to commit?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/ManifestCompactProcedure.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.apache.flink.table.annotation.ArgumentHint;
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.annotation.ProcedureHint;
+import org.apache.flink.table.procedure.ProcedureContext;
+
+import java.util.Collections;
+
+/** Compact manifest file to reduce deleted manifest entries. */
+public class ManifestCompactProcedure extends ProcedureBase {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/ManifestCompactProcedure.java:33]: Also add to spark sql

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CompactManifestProcedure.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.annotation.ProcedureHint;
+import org.apache.flink.table.procedure.ProcedureContext;
+
+import java.util.Collections;
+
+/** Compact manifest file to reduce deleted manifest entries. */
+public class CompactManifestProcedure extends ProcedureBase {
+
+    private static final String commitUser = "Compact-Manifest-Procedure-Committer";
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CompactManifestProcedure.java:35]: 常量【commitUser】命名应全部大写并以下划线分隔

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CompactManifestProcedure.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.apache.flink.table.annotation.ArgumentHint;
+import org.apache.flink.table.annotation.DataTypeHint;
+import org.apache.flink.table.annotation.ProcedureHint;
+import org.apache.flink.table.procedure.ProcedureContext;
+
+import java.util.Collections;
+
+/** Compact manifest file to reduce deleted manifest entries. */
+public class CompactManifestProcedure extends ProcedureBase {
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/procedure/CompactManifestProcedure.java:33]: Why not add the Flink CompactManifestAction?

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommit.java [hunk 6]
```diff
@@ -74,6 +74,8 @@ void overwrite(
 
     void truncateTable(long commitIdentifier);
 
+    void compactManifest();
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommit.java:77]: 接口方法【compactManifest】必须使用javadoc注释

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java (hunk 1 above)
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java]: As a procedure, if you trigger this, maybe merge all the deleted file is the best choise.

**Comment by wwj6591812** [comment]
+1

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java [hunk 7]
```diff
@@ -1104,6 +1064,130 @@ CommitResult tryCommitOnce(
...
+                        latestSnapshot.watermark(),
+                        latestSnapshot.statistics());
+
+        Path newSnapshotPath =
+                branchName.equals(DEFAULT_MAIN_BRANCH)
+                        ? snapshotManager.snapshotPath(newSnapshot.id())
+                        : snapshotManager.copyWithBranch(branchName).snapshotPath(newSnapshot.id());
+
+        if (!commitSnapshotImpl(newSnapshot, newSnapshotPath)) {
+            compactManifest(Pair.of(mergeBeforeManifests, mergeAfterManifests));
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:1147]: Do not use recursion, just use a for loop, and respect commitMaxRetries

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java [hunk 8]
```diff
@@ -1104,6 +1064,130 @@ CommitResult tryCommitOnce(
...
+            }
+        } else {
+            // the fist trial
+            mergeAfterManifests =
+                    ManifestFileMerger.merge(
+                            mergeBeforeManifests,
+                            manifestFile,
+                            manifestTargetSize.getBytes(),
+                            1,
+                            manifestFullCompactionSize.getBytes(),
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:1113]: For the manually compact, we should always do full compaction.

File: paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java [hunk 9]
```diff
@@ -1104,6 +1064,149 @@ CommitResult tryCommitOnce(
...
+                            mergeBeforeManifests,
+                            manifestFile,
+                            manifestTargetSize.getBytes(),
+                            1,
+                            1,
+                            partitionType,
+                            manifestReadParallelism);
+        }
+
+        String baseManifestList = manifestList.write(mergeAfterManifests);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/operation/FileStoreCommitImpl.java:1134]: if mergeAfterManifests is same to mergeBeforeManifests, we can just return.

**Comment by JingsongLi** [comment]
+1

**Pull request merged**